crates/valence_new/src/packet_stream.rs

```diff
@@ -0,0 +1,260 @@
+    pub fn send_drop(&mut self) {
+        self.enc.clear();
+    }
+}
+
+impl WritePacket for PacketStreamer {
+    fn write_packet<P>(&mut self, packet: &P) -> anyhow::Result<()>
+    where
+        P: EncodePacket + ?Sized,
+    {
+        self.try_send(packet)
+    }
+
+    fn write_bytes(&mut self, bytes: &[u8]) -> anyhow::Result<()> {
+        self.send_raw(bytes);
+        Ok(())
+    }
+}
+
+/// Represents a byte stream that packets can be read from and written to.
+pub trait PacketStream {
+    /// Grabs all the bytes buffered for the stream and returns them.
+    fn try_recv(&mut self) -> Result<BytesMut, TryRecvError>;
+
+    /// Writes the given bytes to the stream.
+    fn try_send(&mut self, bytes: BytesMut) -> Result<(), TrySendError>;
+}
+
+pub(crate) struct RealPacketStream {
+    recv: ByteReceiver,
+    send: ByteSender,
+}
+
+impl RealPacketStream {
+    pub(crate) fn new(recv: ByteReceiver, send: ByteSender) -> Self {
+        Self { recv, send }
+    }
+}
+
+impl PacketStream for RealPacketStream {
+    fn try_recv(&mut self) -> Result<BytesMut, TryRecvError> {
+        let bytes = self.recv.try_recv()?;
+        Ok(bytes)
+    }
+
+    fn try_send(&mut self, bytes: BytesMut) -> Result<(), TrySendError> {
+        self.send.try_send(bytes)
+    }
+}
+
+/// A `PacketStream` that reads and writes from an in memory buffer of packets
+/// used for testing.
+pub(crate) struct MockPacketStream {
+    recv_enc: PacketEncoder,
+    pending_recv: BytesMut,
+    send_dec: PacketDecoder,
+    flushed_sent: BytesMut,
+}
+
+impl<'a> MockPacketStream {
+    #[allow(dead_code)]
+    pub(crate) fn new() -> Self {
+        Self {
+            recv_enc: PacketEncoder::new(),
+            pending_recv: BytesMut::new(),
+            send_dec: PacketDecoder::new(),
+            flushed_sent: BytesMut::new(),
+        }
+    }
+
+    /// Clears the sent packets.
+    fn clear_sent(&mut self) {
+        self.flushed_sent.clear();
+    }
+
+    /// Injects a packet into the receive stream as if it were received from a
+    /// client.
+    ///
+    /// ```rust
+    /// use valence_new::packet_stream::MockPacketStream;
+    /// use valence_protocol::packets::c2s::play::KeepAliveC2s;
+    ///
+    /// let mut stream = MockPacketStream::new();
+    /// let packet = KeepAliveC2s { id: 0xdeadbeef };
+    /// stream.inject_recv(packet);
+    /// ```
+    #[allow(dead_code)]
+    pub(crate) fn inject_recv<P>(&mut self, packet: P)
+    where
+        P: EncodePacket,
+    {
+        self.recv_enc
+            .append_packet(&packet)
+            .expect("failed to encode injected packet");
+        let bytes = self.recv_enc.take();
+        self.pending_recv.extend_from_slice(bytes.as_ref());
+    }
+
+    /// Collects all the packets that have been sent so assertions can be made
+    /// on what the server sent in unit tests.
+    #[allow(dead_code)]
+    pub fn collect_sent(&'a mut self) -> anyhow::Result<Vec<S2cPlayPacket<'a>>> {
```

Comment:
So there are a few ways I see to resolve this.
1. Create an entirely new version of `S2cPlayPacket` that owns all of its data. Not fun.
2. Create a new method on `PacketDecoder` that allows you to collect borrowed packets into a vec. I believe a method like this should be possible on `PacketDecoder`
    ```rust
    fn collect_into_vec<'a, P: DecodePacket<'a>>(&'a mut self, vec: &'a mut Vec<P>) -> WhateverResult
    ```
    This isn't great because the method is not very general and would likely require a lot of code duplication to implement. But it can be done.
3. Structure your program in a way that does not require you to collect all the packets into a buffer. Have you considered accepting a list of packets that you expect to match, then walking through that list in order and decoding as you go? It's not a flexible but it might be sufficient for what you're doing.

Reply:
I believe 2 is the best way to go about this. This method would effectively bypass the code paths used in production for compression and encryption. The purpose of setting up this infra is to make assertions on packet contents and their order.  Compression and encryption are already tested, so the omission of these code paths from all other unit tests is acceptable.

Reply:
Alright I've done (2) in https://github.com/valence-rs/valence/commit/04d9588a2d249175428014a31b7e13d5afe501db